controllers/componentdetectionquery_controller.go

```diff
@@ -157,32 +167,144 @@ func (r *ComponentDetectionQueryReconciler) Reconcile(ctx context.Context, req c
+			var devfilesMapReturned map[string][]byte
+			var devfilesURLMapReturned, dockerfileContextMapReturned map[string]string
+			var componentPortsMapReturned map[string][]int
+			revision := source.Revision
+
+			if r.RunKubernetesJob && !(componentDetectionQuery.Annotations["runCDQAnalysisLocal"] == "true") {
+				// perfume cdq job that requires repo cloning and azlier analysis
+				clientset, err := kubernetes.NewForConfig(r.Config)
+				if err != nil {
+					log.Error(err, fmt.Sprintf("Error creating clientset with config... %v", req.NamespacedName))
+					r.SetCompleteConditionAndUpdateCR(ctx, req, &componentDetectionQuery, copiedCDQ, err)
+					return ctrl.Result{}, nil
+				}
+				jobName := req.Name + "-job"
+				var backOffLimit int32 = 0
+				jobSpec := &batchv1.Job{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      jobName,
+						Namespace: req.Namespace,
+					},
+					Spec: batchv1.JobSpec{
+						Template: corev1.PodTemplateSpec{
+							Spec: corev1.PodSpec{
+								ServiceAccountName: "application-service-controller-manager",
+								Containers: []corev1.Container{
+									{
+										Name:            jobName,
+										Image:           r.CdqAnalysisImage,
+										ImagePullPolicy: corev1.PullAlways,
+										Env: []corev1.EnvVar{
+											{
+												Name:  "NAME",
+												Value: req.Name,
+											},
+											{
+												Name:  "NAMESPACE",
+												Value: req.Namespace,
+											},
+											{
+												Name:  "GITHUB_TOKEN",
+												Value: gitToken,
+											},
+											{
+												Name:  "CONTEXT_PATH",
+												Value: context,
+											},
+											{
+												Name:  "REVISION",
+												Value: revision,
+											},
+											{
+												Name:  "URL",
+												Value: source.URL,
+											},
+											{
+												Name:  "DEVFILE_REGISTRY_URL",
+												Value: r.DevfileRegistryURL,
+											},
+											{
+												Name:  "CREATE_K8S_Job",
+												Value: "true",
+											},
+										},
+									},
+								},
+								RestartPolicy: corev1.RestartPolicyNever,
+							},
+						},
+						BackoffLimit: &backOffLimit,
+					},
+				}
+				err = r.Client.Create(ctx, jobSpec, &client.CreateOptions{})
+				if err != nil {
+					log.Error(err, fmt.Sprintf("Error creating cdq analysis job %s... %v", jobName, req.NamespacedName))
+					r.SetCompleteConditionAndUpdateCR(ctx, req, &componentDetectionQuery, copiedCDQ, err)
+					return ctrl.Result{}, nil
+				} else {
+					//print job details
+					log.Info(fmt.Sprintf("Successfully created cdq analysis job %v, waiting for config map to be created... %v", jobName, req.NamespacedName))
+				}
 
-			k8sInfoClient := cdqanalysis.K8sInfoClient{
-				Log:          log,
-				CreateK8sJob: false,
-			}
+				cm, err := waitForConfigMap(clientset, ctx, req.Name, req.Namespace)
```

Comment:
The timeout to wait for the configMap is 5mins but is it generally a lot quicker than that? I see that it's constructed and created in the CDQ module so most of the processing is done - any delays would be attributed to the reading of the configMap from the cluster.

Reply:
yes, generally the CDQ is quicker, with the cdq-image cached on cluster, the entire CDQ time only takes a few seconds. and yes, at this point,most of the processing is done, so we are expecting to use the waiting time as the CDQ timeout time. The reconcile context timeout does not work as expected. it will wait for the sub-process to be done before it checks the timeout. so by switching to use the config map waiting time, we can easily cut the CDQ progress upon timeout and make it under control.

Reply:
and the design for UI side CDQ timeout will be 1min, after 1min, CDQ will still run in backend. we will terminate CDQ if it stuck at waiting for CDQ job to be returned for over 5mins